### ISSUE TITLE
Update dockertools

### ISSUE THREAD

**Issue by dkapitan** [bokeh/bokeh#9581]
@bryevdv @Karel-van-de-Plassche 

Following up on the [discussion how to use Docker for deployments](https://discourse.bokeh.org/t/deploying-bokeh-app-on-google-app-engine-gae-with-docker/4469/10) I am planning to update `docker-tools`.

Looking at the current Dockerfiles, I see `alpine` is used as the base image. Assuming that many Bokeh apps requires `pandas`, and following [these recommendations as to which Docker image is best for Python deployments](https://pythonspeed.com/articles/base-image-python-docker-images/) may plan is to:

- [X] add short explanation on the pro's and cons of different images
- [X] add a table on the various Dockerfiles to summarize their intent (as examples, with different approaches)
~~- [ ] add `Dockerfile-from-pip`, which~~

~~  - is based on `python:3.7-slim` for apps requiring numpy/scipy/pandas~~
~~  - can serve as a template for other Bokeh deployments via `requirements.txt`~~
~~  - uses `demo.bokeh.org` as an example~~
- [x] add `docker-tools/debian/Dockerfile` for building images using `debian-slim` and `conda`
~~- [ ] add `docker-tools/debian/Dockerfile-dev` for building image for development following [the developers guide for getting set up](https://docs.bokeh.org/en/latest/docs/dev_guide/setup.html#devguide-setup)~~
- [x] add a section on how to deploy on Google App Engine

`Dockerfile-from-pip`
```
FROM python:3.7-slim

ENV LANG=C.UTF-8 LC_ALL=C.UTF-8
WORKDIR /some_app

RUN apt-get update \
    && apt-get install gcc -y \
    && apt-get clean
    
COPY . .

RUN pip install --upgrade pip && \
    pip install --no-cache-dir -r requirements.txt

EXPOSE 8080
bokeh serve --port 8080 \
   ...
```

Please let me know what your thoughts are on this, after which I wil submit a PR.

**Comment by Karel-van-de-Plassche** [comment]
I originally chose `alpine`, as it is similar to the systems I'm using and because of the benefits are on the users side (small image footprint), and the downsides mostly on the developers side (have to compile python packages from scratch). I'm using ye old 0.13.0 image to host my own stuff with great success so far. However, I'm really not opposed to switching over to or adding a Debian or Ubuntu-style base image. This lowers the effort on the developers side, with only a tiny cost for the users.

I originally did not go for a ` Dockerfile-from-pip` as to not inflate the amount of images, and conda is the 'recommended' way of using/install Bokeh at the time. Not sure if that changed @bryevdv. Also not opposed to adding the `from-pip` image however.

I'm not sure how many ppl are actually using these images, so far I only know me, @bryevdv, the CI and you.

All suggestions in your list are very good. Go for a PR!

**Comment by bryevdv** [comment]
@dkapitan @Karel-van-de-Plassche thanks for starting this issue. A few notes:

* We have moved entirely off of TravisCI, so these are not even the same VMs that CI uses anymore
* Also, in fact, all the `scripts/ci` bash scripts are gone, so all of that certainly is no longer useful

I will be honest and say that I am not a heavy user of these images, so I don't have many strong opinions here (and I don't really have bandwidth to form them 😄) Off the top of my head ramble of other things that might be helpful:

* For any "dev" images, installing with conda is still the way to go I think (we have tools for obtaining everything needed to run all the tests)
* For other images, I don't really have a pref. i.e. a `Dockerile-from-pip` seems fine with me. 
* Note that the "official" demo site Dockerfile and site is maintained here: https://github.com/bokeh/demo.bokeh.org I don't mind having something similar/duplicate in the main repo though

Basically I'd like to leave things up to you. The only ask I have is to get some best-practices written down for keeping dockerhub up to date. As you can see:

https://hub.docker.com/r/bokeh/bokeh/tags

There is a wildly inconsistent history updates there. I'd really like to eventually including updating dockerhub automatically as part of the normal release process. You guys tell me what needs to be done and I will make sure it starts happening. :)

**Comment by Karel-van-de-Plassche** [comment]
> There is a wildly inconsistent history updates there

Yeah, this is me manually pushing releases that I needed. In principle docker-in-docker has much improved since I wrote these files, so we can probably set up CI to auto-build them on tagged bokeh versions.

**Comment by bryevdv** [comment]
FYI I am currently about to start rebuilding our automation. Due to the new need to compute and include SRI hashes in the built packages we are going to have to have a more multi-staged approach that inserts a layer of testing after packages are built but before they are actually published. I'm trying to collect all the steps and outline the process here: https://github.com/bokeh/bokeh/issues/9341 feel free to suggest where in that it would make sense to add dockerhub automation.

**Comment by dkapitan** [comment]
Just to clarify (since I have little experience in setting up CI): do you guys mean to say that you want automated, new Docker images to be pushed to hub.docker.com, in line with every Bokeh release?

AFAIK the docker images shouldn't change that often, so I don't see what the benefit is of doing this. But I may be missing the point completely, since I also have never used Docker for automated building and/or testing (which is perhaps what you are referring to).

I will go ahead with adding some documentation in any case and will keep you posted.

**Comment by bryevdv** [comment]
@dkapitan I do mean that, but it's easily possible that I am off-base here as well. I as under the impression that pushing with every release version would afford a way for folks to easily get at images for specific versions, which seemed like it might be reasonable to afford.But I don't know if that's needed or if there's actually demand for it though, so again happy tp defer. Are you proposing instead a "generic" image that just always installs the latest version?

**Comment by dkapitan** [comment]
@bryevdv 

> Are you proposing instead a "generic" image that just always installs the latest version?

That's basically my idea for now. To make it easier for users who have just started using Docker, the `Dockerfile-from-pip` would install all Python packages listed in `requirements.txt`, including Bokeh itself. I think this is known territory for the majority of Python users, and they can change that file to their liking. So if they want to use older versions, they can do so, too.

**Comment by bryevdv** [comment]
@dkapitan ok well that's okay with me. I'd suggest if possible a second "dev" image would be helpful. This one would follow the conda-based workflow in the developer's guide Getting Set Up section. The end-result should be that all the tests can be easily run. I think this would help with getting new contributors up and running faster. 

I am pretty vague on how publishing to dockerhub works. I guess we could publish these separately under different names? Should the old images be deleted at some point?

**Comment by dkapitan** [comment]
@bryevdv 
> This one would follow the conda-based workflow in the developer's guide Getting Set Up section.

Ok, that makes sense to me. So for now I will focus on two versions:

- `bokeh-dev` which is based on alpine and conda
- `bokeh-prod` which is based on debian and pip

I will explain the differences between these two in the documentation. For now I propose to upload these manually to [hub.docker.com](https://hub.docker.com).

I propose @Karel-van-de-Plassche remains maintainer/contact for the alpine-based images. I will put up my name for the debian version.

**Comment by Karel-van-de-Plassche** [comment]
If you can check the final size difference between the alpine-based image and the debain-based image, I might be good to move everything to debian, to lower maintenance burden a bit.

>  I as under the impression that pushing with every release version would afford a way for folks to easily get at images for specific versions, which seemed like it might be reasonable to afford.

This was my use case for the Docker images at least. People like me have only little time to keep up with API changes unfortunately, so 'freezing' bokeh in a container is very convenient for me. As I/we did not get many comments on the Docker images, I think it's only a small group of users that require this.

**Comment by dkapitan** [comment]
@Karel-van-de-Plassche @bryevdv 

> If you can check the final size difference between the alpine-based image and the debain-based image, I might be good to move everything to debian, to lower maintenance burden a bit.

Difference between [alpine-miniconda3](https://hub.docker.com/r/frolvlad/alpine-miniconda3/tags) and [python:3.7-slim](https://hub.docker.com/_/python?tab=tags&page=1&name=3.7-slim) is neglible: there are both around 50.5 MB compressed. So I am in favour of moving everything over to Debian.

If we do that, than it's just a choice of:

* which package manager, i.e. pip vs conda, people prefer. I guess that's just personal preference unless the Bokeh community has a clear preference for conda
* whether you want to build from source or not, probably dependent on whether you are developing Bokeh or not

So please let me know if we want to go Debian all the way.

**Comment by dkapitan** [comment]
This seems relevant: https://jcrist.github.io/conda-docker-tips.html

**Comment by dkapitan** [comment]
After reading [Jake's article on conda vs pip](https://jakevdp.github.io/blog/2016/08/25/conda-myths-and-misconceptions/) am inclined to stick to conda.

**Comment by dkapitan** [comment]
@bryevdv @Karel-van-de-Plassche 

Hmm, getting a bit confused now. Have build the miniconda3 images from continuumio and getting huge difference in image size.

Not sure what's causing this?

```
REPOSITORY                   TAG                 IMAGE ID            CREATED             SIZE
dkapitan/miniconda3-alpine   latest              c4a16cf7a0ba        6 minutes ago       163MB
dkapitan/miniconda3          latest              d10f7fd894d9        10 minutes ago      430MB
```

**Comment by Karel-van-de-Plassche** [comment]
In general a miniconda/anaconda/conda distribution is way larger than a 'bare' python distribution. So maybe you were comparing debian + bare python vs alpine + miniconda before? If we care about size, you can try and strip the conda image down, e.g. https://jcrist.github.io/conda-docker-tips.html

**Comment by dkapitan** [comment]
@bryevdv @Karel-van-de-Plassche 

So I have done a short comparison of different images from https://hub.docker.com which seemed legit to me. My aim is to get a feel of the base Python 3.7 image that already has the compiled non-Python libraries installed (uncompressed image size). 

|  | Alpine | Debian slim |
|:---|:---|:---|
|pip | [jfloff/alpine-python:3.7-slim](https://hub.docker.com/r/jfloff/alpine-python): 82 MB | [mangothecat/pydata-pip](https://hub.docker.com/r/mangothecat/pydata-pip) 154 MB |
|conda |[continuumio/miniconda3:alpine from GitHub](https://github.com/ContinuumIO/docker-images/blob/master/miniconda3/alpine/Dockerfile) 163 MB | [continuumio/miniconda3:debian-slim from GitHub](https://github.com/ContinuumIO/docker-images/blob/master/miniconda3/debian/Dockerfile) 245 MB |

So there is a difference. However, given that a modest Bokeh app (like demo.bokeh.org) quickly adds up to an image size of around 1 GB (uncompressed), the difference for production systems is just 10-15%. Note that these conda images already have the optimisations mentioned by https://jcrist.github.io/conda-docker-tips.html.

For me, the most important trade-offs are:

* maintainability of both Python and non-Python dependencies --> conda seems the better choice
* stability of non-Python dependencies --> I have read some disconcerting stories about using Alpine with Python, but have no firsthand experience.

My idea is to add these findings in the introduction of bokeh/docker-tools, so users can make an informed decision themselves.

Question that stills piques me, is to what extent Alpine is sufficiently stable for Python apps that required the pydata/numpy stack. Interested to hear your thoughts and experience on this.

**Comment by Karel-van-de-Plassche** [comment]
> Question that stills piques me, is to what extent Alpine is sufficiently stable for Python apps that required the pydata/numpy stack. Interested to hear your thoughts and experience on this.

In principle Alpine is stable enough. It's based on Arch linux, which even though it is bleeding edge, I use as my daily system without problems for the past 4 years or so. However, important is that python package need to be recompiled by pip on install, as there are no wheels available. This does mean significant image build/debugging time for the developer. For easy-of-the-developer, maybe a debian/ubuntu-based approach is better. I would try to avoid maintaining both an Alpine and Debian image, so in that case I'd take the image size hit and choose Debian.

**Comment by dkapitan** [comment]
@Karel-van-de-Plassche
Thanks for the info. Will indeed stick to Debian for now. Incidentally, my understanding is that Alpine Linux is based on Busybox, and not Arch Linux.

**Comment by Karel-van-de-Plassche** [comment]
@dkapitan 

You are completely right:

> Alpine Linux is a security-oriented, lightweight Linux distribution based on musl libc and busybox.

It uses apk for package management

**Comment by dkapitan** [comment]
@Karel-van-de-Plassche @bryevdv 
have started working on this issue and am testing whether everything work. `alpine/Dockerfile-from-conda` fails with `ValueError: setting 'version' makes sense only when 'mode' is set to 'cdn'
as per [GH30](https://github.com/bokeh/jupyter_bokeh/issues/30). Any suggestions as to how to fix this?

**Comment by bryevdv** [comment]
@dkapitan what is setting `BOKEH_VERSION`? Can you make it not do that? Otherwise I suggested some potential changes to Bokeh that could be made (but no one ever expressed any interest to make them)

**Comment by bryevdv** [comment]
I should add, we could consider changing  that variable to something else, like `BOKEH_CDN_VERSION` that might be less likely to clash. But that change would need to go in *now* for 2.0 since it is a breaking change.

**Comment by bryevdv** [comment]
I've made https://github.com/bokeh/bokeh/pull/9713

**Comment by dkapitan** [comment]
> @dkapitan what is setting `BOKEH_VERSION`? Can you make it not do that? Otherwise I suggested some potential changes to Bokeh that could be made (but no one ever expressed any interest to make them)

@bryevdv this is set in the current Dockerfiles in`docker-tools`. It has been put there to allow building older versions. I am happy to get rid of those, i.e. focusing on a new version of `docker-tools` that builds the latest version.

**Comment by dkapitan** [comment]
> I should add, we could consider changing that variable to something else, like `BOKEH_CDN_VERSION` that might be less likely to clash. But that change would need to go in _now_ for 2.0 since it is a breaking change.

@bryevdv 
Have commented-out `BOKEH_VERSION` setting for now, and it works. Have just published my first PR for bokeh (yay!), which includes everything from this issue apart from the development image. That will take more time for me to get my head around, so thought it was better to submit what I have now. Looking forward to your feedback.

**Comment by Karel-van-de-Plassche** [comment]
`BOKEH_VERSION` was meant to be able to retroactively build images for older bokeh versions if users are requesting those. Seeing there has not been a request for the past few years, maybe not worth maintaining this ;)

**Comment by dkapitan** [comment]
Have openend new issue for adding development docker image #9721. This issue can be closed once PR has been accepted.
